src/MagnumPlugins/WebPImporter/Test/WebPImporterTest.cpp

```diff
@@ -0,0 +1,411 @@
+        /* Explicitly checking non-null but empty view */
+        case 0:
+            CORRADE_VERIFY(!importer->openData({data.data, 0}));
+            break;
+        /* Checking a 0-byte webp file */
+        case 1:
+            CORRADE_VERIFY(!importer->openFile(Utility::Path::join(WEBPIMPORTER_TEST_DIR, data.filename)));
+            break;
+    }
+    CORRADE_COMPARE(out.str(), "Trade::WebPImporter::openData(): the file is empty\n");
+}
+
+void WebPImporterTest::invalid() {
+    auto&& data = InvalidData[testCaseInstanceId()];
+    setTestCaseDescription(data.name);
+
+    Containers::Pointer<AbstractImporter> importer = _manager.instantiate("WebPImporter");
+
+    if(testCaseInstanceId() == 0) {
+        /* The open does just a memory copy, so it doesn't fail */
+        CORRADE_VERIFY(importer->openData(data.data.prefix(data.data.size() - 1)));
+    } else {
+        CORRADE_VERIFY(importer->openFile(Utility::Path::join(WEBPIMPORTER_TEST_DIR, data.filename)));
+    }
+    std::ostringstream out;
+    Error redirectError{&out};
+    CORRADE_VERIFY(!importer->image2D(0));
+    CORRADE_COMPARE(out.str(), data.error);
+}
+
+void WebPImporterTest::rgb() {
+    auto&& data = RgbData[testCaseInstanceId()];
+    setTestCaseDescription(data.name);
+
+    Containers::Pointer<AbstractImporter> importer = _manager.instantiate("WebPImporter");
+    CORRADE_VERIFY(importer->openFile(Utility::Path::join(WEBPIMPORTER_TEST_DIR, data.filename)));
+
+    Containers::Optional<Trade::ImageData2D> image = importer->image2D(0);
+    CORRADE_VERIFY(image);
+    CORRADE_COMPARE(image->size(), Vector2i(3, 3));
+    CORRADE_COMPARE(image->format(), PixelFormat::RGB8Unorm);
+
+    /* The image has four-byte aligned rows, clear the padding to deterministic
+       values */
+    CORRADE_COMPARE(image->data().size(), 36);
+    image->mutableData()[9] = image->mutableData()[10] =
+        image->mutableData()[11] = image->mutableData()[21] =
+            image->mutableData()[22] = image->mutableData()[23] =
+                image->mutableData()[33] = image->mutableData()[34] =
+                        image->mutableData()[35] = 0;
+    switch (testCaseInstanceId()) {
+        /* Lossless bitstream */
+        case 0:
+            CORRADE_COMPARE_AS(image->data(), Containers::arrayView<char>({
+                '\x1e', '\x6e', '\x1e',
+                '\x1e', '\x6e', '\x1e',
+                '\x1e', '\x6e', '\x1e', 0, 0, 0,
+
+                '\xef', '\x91', '\x91',
+                '\xef', '\x91', '\x91',
+                '\xef', '\x91', '\x91', 0, 0, 0,
+
+                '\x52', '\x52', '\xbe',
+                '\x52', '\x52', '\xbe',
+                '\x52', '\x52', '\xbe', 0, 0, 0,
+            }), TestSuite::Compare::Container);
+            break;
+
+        /* Same image with 90% image quality (lossy bitstream) */
+        case 1:
+            CORRADE_COMPARE_AS(image->data(), Containers::arrayView<char>({
+                '\x4d', '\x55', '\x24',
```

Comment:
Testing for bit-exact equality of lossy-compressed images is quite a headache .. as you probably already found out :) And it's not unlikely that the output would change slightly when a new WebP version comes out.

So let's try a fuzzy comparison using [DebugTools::CompareImage](https://doc.magnum.graphics/magnum/classMagnum_1_1DebugTools_1_1CompareImage.html) instead, where you can say how much can the pixels differ from the lossless ground truth. Those thresholds are best put into the instance data to not have to branch on `testCaseInstanceId()` again:

```cpp
constexpr struct {
    const char* name;
    const char* filename;
    Float maxThreshold, meanThreshold;
} RgbData[]{
    {"lossless", "rgb-lossless.webp", 0.0f, 0.0f},
    {"lossy with 90% image quality", "rgb-lossy-90.webp", 10.0f, 2.0f},
};
```

The lossless image should be *exactly* the pixel values we expect, which is why the zeros. For the lossy I picked `10.0f` and `2.0f` just at random, when you run the test it'll tell you what's the actual threshold and a (slightly larger) value is then what should go here instead.

I also don't think we need to test that many variants, so just a losless and the 90% quality could suffice -- again, our goal is to just make sure the library is used correctly by the plugin, it's the WebP authors' job to ensure the library gives reasonable output for various quality settings.

The test would then look like this, I hope it's not much of a handholding this way :) Since it uses the DebugTools library now, you have to find it in CMake and link the test to it (look into `BasisImporter/Test/CMakeLists.txt`, there it's also being used). 

```cpp
#include <Magnum/DebugTools/CompareImage.h>

…

void WebPImporterTest::rgb() {
    auto&& data = RgbData[testCaseInstanceId()];
    setTestCaseDescription(data.name);

    Containers::Pointer<AbstractImporter> importer = _manager.instantiate("WebPImporter");
    CORRADE_VERIFY(importer->openFile(Utility::Path::join(WEBPIMPORTER_TEST_DIR, data.filename)));

    Containers::Optional<Trade::ImageData2D> image = importer->image2D(0);
    CORRADE_VERIFY(image);
    /* Format and size checked by CompareImage, the image should also have 
       four-byte aligned rows */
    CORRADE_COMPARE(image->storage().alignment(), 4);
    
    const char expected[]{
        '\x1e', '\x6e', '\x1e',
        '\x1e', '\x6e', '\x1e',
        '\x1e', '\x6e', '\x1e', 0, 0, 0,

        '\xef', '\x91', '\x91',
        '\xef', '\x91', '\x91',
        '\xef', '\x91', '\x91', 0, 0, 0,

        '\x52', '\x52', '\xbe',
        '\x52', '\x52', '\xbe',
        '\x52', '\x52', '\xbe', 0, 0, 0,
    };
    
    CORRADE_COMPARE_WITH(*image, 
        (ImageView2D{PixelFormat::RGB8Unorm, {3, 3}, expected}),
        (DebugTools::CompareImage{data.maxThreshold, data.meanThreshold}));
}
```

And similar for the RGBA test.

Reply:
This definitely looks much cleaner and better :) 

> when you run the test it'll tell you what's the actual threshold and a (slightly larger) value is then what should go here instead.

In my first commit after determining the threshold, I got CI errors because the values were not large enough for other some Linux versions. Hence, I increased them a bit more, now they are all 0.2 larger than actual threshold.

Reply:
> the values were not large enough for other some Linux versions

That's quite interesting, and telling something about this format. When I said "it's not unlikely that the output would change slightly when a new WebP version comes out" above, I didn't expect that to happen so soon and so dramatically that it breaks even fuzzy comparison :D Other lossy codecs such as JPEG put a lot of effort into ensuring that the decoded output of the same file is bit-exact across versions, and if it changes, it's viewed as a critical bug. A bit sad to see that WebP doesn't live up to that standard. 

It's like a Butterfly Effect, even a single-bit change can result in large differences down the line if the image is further recompressed or if it's used as a source for texture baking operations. And then a minor game patch that was meant just to improve image loading times is suddenly several gigabytes large.

Reply:
Yes, I thought it was interesting but had no idea about its consequences. Could it be prevented by forcing the installation of a specific WebP version and check via CMake configuration if the correct version is installed?

Reply:
Possibly, but staying at a fixed version has other downsides such as unpatched security issues. More like just something to keep in mind, that unexplainable minor differences in rendered/baked output might be due to a WebP upgrade. A radical suggestion might be to not store your source assets in WebP and treat it only as an export format on which nothing else depends.

Vaguely related -- I came across a similar issue when upgrading a 3rd party glTF parser once, somehow it started parsing transformations slightly differently and [led to unexpected test failures](https://github.com/facebookresearch/habitat-sim/pull/496#issuecomment-587461427). Wasted quite some time trying to find the root cause.